vo: change vsync base to nanoseconds

kasper93 opened the pull request.

Dudemanguy commented:
UST is actually "Unadjusted System Time". It has nothing to do with microseconds (i.e. please don't change that).

kasper93 commented:
> UST is actually "Unadjusted System Time". It has nothing to do with microseconds (i.e. please don't change that).

Ok, fixed. My bad about the name change. For drm I reverted all changes, because apparently it works with microsecond timestamps. wayland and x11 goes through mpv's `present_sync` so it is fine to not truncate the value in case of wayland. I don't know what resolution is in `XPresentCompleteNotifyEvent`, but it was working with microseconds before so I assume that.

sfan5 commented on audio/out/ao_audiotrack.c:
```diff
@@ -372,26 +372,26 @@ static uint32_t AudioTrack_getPlaybackHeadPosition(struct ao *ao)
-                if (utime1 != utime2 && utime2 != 0 && fpos != 0) {
+                int64_t nanotime2 = MP_JNI_GET_LONG(p->timestamp, AudioTimestamp.nanoTime);
+                //MP_VERBOSE(ao, "getTimestamp: fpos= %u / time= %"PRId64" / now= %"PRId64" / stable= %d\n", fpos, nanotime2, now, p->timestamp_stable);
+                if (nanotime1 != nanotime2 && nanotime2 != 0 && fpos != 0) {
```
what do these changes have to do with the timer code?

also a bit lower the calculations still work with microseconds

Edit: I noticed you removed `mp_raw_time_us`, I think this code can just use `mp_time_us`.

kasper93 replied:
We get ns timestamp, see `MP_JNI_GET_LONG(p->timestamp, AudioTimestamp.nanoTime)`, no need to keep truncating after replacing `mp_raw_time_us` with `mp_raw_time_ns`, it is jsut related change to that.

sfan5 replied:
Sure but this code just below still uses microseconds:
![grafik](https://github.com/mpv-player/mpv/assets/1042418/dc86455d-91e5-4707-b438-e52ce2e0771b)

I'd rather have as few as possible things changed in this file or all code checked and changed to use ns at once.
Since this isn't even in scope for the PR I prefer the former.

sfan5 commented on osdep/timer-linux.c:
```diff
@@ -34,22 +32,16 @@ void mp_sleep_us(int64_t us)
-uint64_t mp_raw_time_us(void)
-{
-    struct timeval tv;
-    gettimeofday(&tv,NULL);
```
"Using gettimeofday() as a precise timepoint is not that good idea."
but as you can see here we're not doing that, it's merely a fallback

kasper93 replied:
Fair, will reword this commit message. We are still truncating,

sfan5 commented on osdep/timer-linux.c:
```diff
@@ -34,22 +32,16 @@ void mp_sleep_us(int64_t us)
-    gettimeofday(&tv,NULL);
-    return tv.tv_sec * 1000000LL + tv.tv_usec;
-}
+    timespec_get(&tp, TIME_UTC);
```
I don't have a manpage for this function on my system, where does it come from?

kasper93 replied:
C11, https://en.cppreference.com/w/c/chrono/timespec_get

sfan5 commented on osdep/timer.c:
```diff
@@ -49,15 +49,20 @@ void mp_time_init(void)
 double mp_time_sec(void)
 {
-    return mp_time_us() / (double)(1000 * 1000);
+    return mp_time_ns() / 1e9;
```
is `1e9` double-typed or will this do an integer division?

kasper93 replied:
Yes, `1e9` is double literal.

sfan5 commented on video/out/present_sync.c:
```diff
@@ -71,8 +71,8 @@ void present_sync_swap(struct mp_present *present)
-    uint64_t now_monotonic = ts.tv_sec * 1000000LL + ts.tv_nsec / 1000;
-    uint64_t ust_mp_time = mp_time_us() - (now_monotonic - ust);
+    uint64_t now_monotonic = ts.tv_sec * UINT64_C(1000000000) + ts.tv_nsec;
+    uint64_t ust_mp_time = mp_time_ns() - (now_monotonic - ust);
```
this changes the assumption on what unit `ust` is from microseconds to nanoseconds. intended?

kasper93 replied:
This is fully part of extending vsync precision. `present_sync` is fully controlled by us and is feed from wayland/x11 presentation feedback. `ust` here is defined by mpv and it is this commit goal to change it to ns. Can't really split it up without breaking things.

Dudemanguy replied:
I was wrong earlier, the 1000 multiplication you did for xorg was correct. I did some reading and wayland's presentation time nanosecond part is extra. The UST part (tv_sec_lo + tv_sec_hi) is what's equivalent to xorg and I guess it is microseconds. Or well that's what we've been assuming at least it works.

Dudemanguy replied:
Actually, I realized that converting `ust` to ns breaks a lot of other assumptions. e.g. wayland uses the calculated vsync_duration value and the `ust_passed/msc_passed` part only works if ust is microseconds.

kasper93 replied:
Yes, I was testing on Wayland briefly today and it like you say doesn't work. This patch is half baked. Will need to properly test everything. `vsync_duration` is fine actually, the other one is messed up. Don't worry about it for now, will do proper patch.

Dudemanguy replied:
You could just leave the present sync stuff in microseconds. That's basically what you did for drm and it works there. Honestly, there's no reason drm can't also use the present_sync codepath. I'm pretty sure the sbc stuff useless, but that's something for another day.

sfan5 replied:
on x11 `ust` comes directly from `XPresentCompleteNotifyEvent` and you did not modify that part of code.
so either:
1. the new assumption is that x11 provides microseconds
2. this change here is incorrect

sfan5 commented:
What issue is that actually meant to solve? I don't see anything concrete mentioned anywhere.

kasper93 commented:
> What issue is that actually meant to solve? I don't see anything concrete mentioned anywhere.

Same as https://github.com/mpv-player/mpv/commit/bda84399ab363d3ac5e79fcf6eaa40fcb60f32ae fixed truncation of `nominal_vsync_interval`  by changing it to double, this commit fixes the same truncation when calculating `estimated_vsync_interval`. But it is more principal way. We get at least ns resolution from Wayland/X11/D3D11 feedback, there is no need to truncate this. Which is historical baggage, because our timers were maxed at microseconds. While all platforms mpv targets provide higher resolution timers, there is literally no need to truncate this.

Also this is a base for further improving sync code, but first I want to have at least native timer resolution and not dealing with truncated values. Again same fix as https://github.com/mpv-player/mpv/commit/bda84399ab363d3ac5e79fcf6eaa40fcb60f32ae but with broader scope.

I'm traveling this week, so will not have much time to improve/test it. But if someone wants to try feel free.

I'm mostly thinking that we should have timer sync logic, to infer clock base of UST instead of assuming it's base, because it doesn't seem to be in fact defined. But baby steps, I want to incrementally work on this.

Dudemanguy commented:
The presentation stuff does seem a bit messed up (wayland and x11). Honestly just looking at it I'm not sure exactly why. I guess maybe a missed conversion somewhere.

Dudemanguy commented on video/out/wayland_common.c:
```diff
@@ -2581,10 +2581,10 @@ void vo_wayland_wait_frame(struct vo_wayland_state *wl)
     // Completely arbitrary amount of additional time to wait.
     vblank_time += 0.05 * vblank_time;
-    int64_t finish_time = mp_time_us() + vblank_time;
+    int64_t finish_time = mp_time_ns() + vblank_time;
```
`vblank_time` is in microseconds.

sfan5 commented:
> Same as bda8439 fixed truncation of `nominal_vsync_interval` by changing it to double, this commit fixes the same truncation when calculating `estimated_vsync_interval`.

Alright, for some reason I thought that was different. Indeed the same 16666(.66666) truncation applies when you use microseconds over a smaller unit or floating-point.

kasper93 commented on osdep/timer-win2.c:
```diff
@@ -66,16 +66,16 @@ void mp_sleep_us(int64_t us)
     LARGE_INTEGER perf_count;
     QueryPerformanceCounter(&perf_count);
 
     // Convert QPC units (1/perf_freq seconds) to microseconds. This will work
```
We are fine. We have plenty of years to hold in int64_t :)

```
>>> ((1 << 63) - 1) / 1e9 / 60 / 60 / 24 / 365
292.471208677536
```

kasper93 commented:
@Dudemanguy: So the main issue, was the I assumption, same as the code currently assumes, that `feedback_presented` gives us presentation feedback for the past events. And I actually wraparound uint64_t value on the subtraction. 

You can reproduce this on master adding following code to the `feedback_presented` function. 
```
    struct timespec ts;
    if (clock_gettime(CLOCK_MONOTONIC, &ts)) {
        return;
    }
    uint64_t now_monotonic = ts.tv_sec * UINT64_C(1000000) + ts.tv_nsec / 1000;
    assert(now_monotonic >= ust);
```

I reviewed sync code in general and made few notes of things that I'd like to change. I have vague plan in head, but this PR is not it at this moment.

Dudemanguy commented:
I actually noticed that happening as well the other day. I'm not sure how we should interpret such a value; it reeks of nonsense honestly.

kasper93 commented:
I need to read docs how exactly it is defined for wayland. But on Windows `DwmGetCompositionTimingInfo` the `qpcVBlank` field is most of the time in the future. I think once it is scheduled the value is already available. But I don't want to think about Windows specifics rn....

Dudemanguy commented:
On another look, the changes LGTM. Works fine too. The only thing is the mac stuff which I'm not sure about. The README claims we support macOS 10.8 (no idea if this is even true), but clock_gettime apparently requires 10.12. I doubt we really care about ancient mac version though.

Dudemanguy commented on meson.build:
```diff
@@ -364,6 +362,11 @@ if features['win32-internal-pthreads']
+    # pthreads is intentionally left undefined in win32 branch to find incorrect
+    # uses of it immediately
+    pthreads = dependency('threads')
+    dependencies += pthreads
```
Does this eliminate the need for the workaround described above (putting `HAVE_WIN32_INTERNAL_PTHREADS`in `common/stats.c`)?

kasper93 replied:
Not really, because apparently it is pulled from, `pthread_time.h` which is got pulled from `time.h`. But to make it clearer we can defined `#define _POSIX_TIMERS 0` in our wrapper header.

I don't like how we use the same name for a header `pthread.h` instead of adding `mp_thread.h` or something and include what's needed. But well it works as is.

Dudemanguy replied:
Feel free to change it if you want. When I added meson support I didn't want to mess with existing files too much since waf existed, but that's no concern anymore.

kasper93 commented:
> The README claims we support macOS 10.8 (no idea if this is even true), but clock_gettime apparently requires 10.12. I doubt we really care about ancient mac versions though.

Ok, I reverted this change. I thought 10.12 is old enough, but since we explicit say 10.8 in readme, let's keep it.

Dudemanguy reviewed (approved):
Can't speak for the android or mac changes, but LGTM.

avih commented:
Even if we agree that nanosec is crucial for vsync (which I doubt, but whatever), why does this PR also change everything else which uses timing to nanosec?

usec was enough for everything else, so just leave it usec everywhere and change it to nanosec only where nanosec solves a problem.

The existing usec functions can now be implemented diiferently, if required, but everything which uses them can stay unmodified.

Dudemanguy commented:
>why does this PR also change everything else which uses timing to nanosec?

It doesn't. I don't know why think that. Just grep `mp_time_us`. This only touches the tip of the iceberg honestly (not trying to imply that everything needs to be changed).

>Also, the change of the name mp_add_timeout to mp_time_us_add is probably bad, because the name became a lot less obvious about what it does IMO.

Why? There's no actual timeout. It literally just adds more time to a value.

Dudemanguy commented on video/out/vo.c:
```diff
@@ -764,7 +764,7 @@ void vo_wakeup(struct vo *vo)
 {
     struct vo_internal *in = vo->in;
-    int64_t now = mp_time_us();
+    int64_t now = mp_time_ns();
```
Woops missed this one. Is this right? Shouldn't duration and so on still be in microseconds?

kasper93 replied:
Looks like a stray change indeed, it is not even related to anything I was working on. Fixed.

avih commented:
Nevermind.

I still doubt that nanosec accuracy is required because the measuring itself is noisy to begin with, nano or u sec, so accounting for that noise is needed either way, and I doubt that accounting for usec noise would yield worse results than accounting for nanosec noise, but whatever.

kasper93 commented:
> I still doubt that nanosec accuracy is required

The original change was initiated by people complaining that `1 / 60` in usec is not accurate enough.

> because the measuring itself is noisy to begin with, nano or u sec, so accounting for that noise is needed either way, and I doubt that accounting for usec noise would yield worse results than accounting for nanosec noise

We are accounting for the noise, by averaging 200 past samples. Averaging truncated values introduces more error/bias. And adding more precision is free.

> but whatever.

Yes, whatever. Thanks. I really don't understand this mentality, of not touching/improving things at all cost. Note that you did not make the changes, I did, so whatever...

Wayland gives UST in nanoseconds, Windows uses QPC for their feedback. There is literally no gain in truncating those values, just because. Same in other places where we have higher precision timebase. Nowadays all platforms provides high resolution clocks we can utilize them. It is free. It doesn't harm anything to store more digits in int64_t, but whatever.

Dudemanguy commented:
There's several improvements to the vsync code that could be made. I have a few ideas in mind. But step one would be using higher precision clocks so I'm all for this. Actually I plan to follow up with some more unit changes after this PR (unless kasper beats me too it hah).

kasper93 commented:
@Dudemanguy: Go ahead, I don't have anything in the pipeline right now. I've seen some places that would benefit form higher precision timestamps though.

I want to refactor how win32 presentation feedback works, but I think we are not in conflict there :)

Dudemanguy commented on osdep/timer.c:
```diff
@@ -103,8 +103,6 @@ struct timespec mp_time_us_to_realtime(int64_t time_us)
         diff_nsecs -= 1000000000UL;
     }
-    // OSX can't deal with large timeouts. Also handles tv_sec/time_t overflows.
-    diff_secs = MPMIN(diff_secs, 10000000);
```
macOS is apparently such a garbage platform that we still need this. How about just guarding it with `#if HAVE_DARWIN` instead.
https://github.com/dicej/ipmpsc/issues/4

Akemi replied:
the hate isn't really necessary like this, or do we want to start another gnome situation?

kasper93 replied:
@Dudemanguy: In fact, I reintroduced this later, set to 1000 days... to guard for time_t overflow. We call wait function with infinite time in few places, so this has to be clamped anyway. I hope this will be enough also for OSX.

@Akemi: Don't worry winpthreads is even more garbage it cannot handle wait times that are close less than current time, but also not more than 1s...

Dudemanguy replied:
Everything sucks. Well I guess we can just try it and see what happens.

kasper93 replied:
I should probably rebase and remove this commit. Was too lazy to do that.

kasper93 replied:
Done now...

kasper93 replied:
We really should replace things like `wait_wakeup(ctx, INT64_MAX);` to call `pthread_cond_wait` instead. But it is hard to track all of them, when the wait time comes from variable.

Dudemanguy commented on osdep/timer-win2.c:
```diff
@@ -66,16 +66,16 @@ void mp_sleep_us(int64_t us)
     LARGE_INTEGER perf_count;
     QueryPerformanceCounter(&perf_count);
 
     // Convert QPC units (1/perf_freq seconds) to microseconds. This will work
```
Just noticed but this should say `nanoseconds` now.

kasper93 replied:
fixed

Dudemanguy commented:
Let's see if someone catches something.
